autode/opt/coordinates/dic.py

```diff
@@ -355,88 +375,84 @@ def active_indexes(self) -> List[int]:
 
         return [i for i in range(n + m) if i not in self.inactive_indexes]
 
-    def _update_g_from_cart_g(self, arr: Optional["Gradient"]) -> None:
-        r"""
-        Updates the gradient from a calculated Cartesian gradient, where
-        the gradient is that of the Lagrangian. Includes dL/d_λi terms where
-        λi is the i-th lagrangian multiplier.
-
-        -----------------------------------------------------------------------
-        Arguments:
-            arr: Cartesian gradient array
+    def iadd(self, value: np.ndarray) -> "OptCoordinates":
         """
+        Add a step in internal coordinates (along with Lagrange multipliers)
+        to this set of coordinates, and update the Cartesian coordinates
 
-        if arr is None:
-            self._x.g, self.g = None, None
-            return
-
-        assert self._lambda is not None, "Must have λ defined"
+        Args:
+            value: Difference between current and new DICs, and the multipliers
+        """
+        assert len(value) == len(self) + self.n_constraints
+        # separate the coordinates and the lagrange multipliers
+        if self.n_constraints > 0:
+            delta_lambda = value[-self.n_constraints :]
+            self._lambda += delta_lambda
+            delta_s = value[: -self.n_constraints]
+        else:
+            delta_s = value
 
-        self._x.g = arr.flatten()
-        n = len(self)
-        m = self.n_constraints
+        return super().iadd(delta_s)
 
-        self.g = np.zeros(shape=(n + m,))
+    @property
+    def g(self):
+        """
+        Gradient of the energy, contains the Lagrangian dL/d_λi terms where
+        λi is the i-th lagrangian multiplier.
+        """
+        if self._g is None:
+            return None
 
-        # Set the first part dL/ds_i
-        self.g[:n] = np.matmul(self.B_T_inv.T, self._x.g)
+        n, m = len(self), self.n_constraints
+        arr = np.zeros(shape=(n + m,))
+        arr[:n] = self._g
```

Comment:
`self._g` and `self.g` have different shapes? this seems v confusing

Reply:
@t-young31 I am not really sure what is the best way to put the lagrangian terms into gradient. The original code irreversibly modified the gradient and assigned to `self._g`. But then it's not really possible to update the molecular Hessian with the unmodified gradient. The CRFO paper says that the molecular hessian should be updated, but before the whole Lagrangian hessian was updated. I am not sure of the pros and cons of either approach - but just wanted to follow the paper (one thing also is that the molecular hessian should be close to positive definite - but the Lagrangian hessian won't be as the constraint modes would have negative eigenvalues at convergence so BFGS would struggle with the full Lagrangian Hessian).

Reply:
I am _fairly_ sure I tried both when I wrote this initially, but definitely worth testing out. agree that BFGS should have a tough time updating the whole thing. perhaps worth trying a Bofill update too? 
It was ages ago so I could be completely wrong of course!

Reply:
@t-young31 Ok, I tested this on a small Pd complex with two constraints - bond(0-1) set to 1.830 A and bond(0-5) set to 2.411 A. These are the optimisation trajectories

[twoconstr-bfgssr1.trj.xyz.txt](https://github.com/user-attachments/files/15986009/twoconstr-bfgssr1.trj.xyz.txt)

[twoconstr-bofill.trj.xyz.txt](https://github.com/user-attachments/files/15986006/twoconstr-bofill.trj.xyz.txt)

[twoconstr-molec-bfgssr1.trj.xyz.txt](https://github.com/user-attachments/files/15986004/twoconstr-molec-bfgssr1.trj.xyz.txt)

with the original code (updating the total Lagrangian hessian with BFGS-SR1) it takes 12 iters to converge. With Bofill update and the original code it does not converge in 20 iters. With updating only the molecular hessian with BFGS-SR1 it takes 4 iters to converge.

What is interesting is that with BFGS-SR1 update of the Lagrangian, the constraints take 9 iters to satisfy, with Bofill it takes 4 iters to satisfy (maybe because it is good with negative eigenvalues in hessian). But minimisation is very slow with Bofill so it does not converge even after the constraints are satisfied and eliminated from the space.

Reply:
I think updating the molecular hessian is probably better. With updating the lagrangian - there are a lot of jitters in the trajectory not sure why.

Reply:
Would it be better if the 'self.g' setter was removed entirely from all classes? i.e. gradients can only be set by `update_g_from_cart_g()` and only accessed by `self.g` (the `self._g` attribute is hidden and not supposed to be accessed by users in any case)

Reply:
awesome – molecular update it is

I'm not sure there's too much to gain by removing the public `g` but don't have strong feelings on the matter. happy to roll with the complexity of the different shapes